src/formats/base_32.cpp

```diff
@@ -0,0 +1,187 @@
+    }
+    return result;
+}
+
+bool verify_checksum_base32(const std::string& hr_part, const data_chunk& data_part)
+{
+    data_chunk expanded = hrp_expand_base32(hr_part);
+    expanded.insert(expanded.end(), data_part.begin(), data_part.end());
+    return polymod_base32(expanded) == 1;
+}
+
+data_chunk create_checksum_base32(const std::string& hr_part, const data_chunk& data_part)
+{
+    data_chunk values = hrp_expand_base32(hr_part);
+    values.insert(values.begin(), data_part.begin(), data_part.end());
+    values.resize(values.size() + checksum_length);
+    uint32_t polymod = polymod_base32(values) ^ 1;
+    data_chunk result(checksum_length, 0);
+    for (size_t i = 0; i < result.size(); ++i)
+    {
+        result[i] = (polymod >> (5 * (5 - i))) & 31;
+    }
+    return result;
+}
+
+std::string encode_base32(const std::string& hr_part, const data_chunk& data_part)
+{
+    std::string encoded = hr_part + separator;
+    encoded.reserve(encoded.size() + data_part.size() + checksum_length);
+    data_chunk checksum = create_checksum_base32(hr_part, data_part);
+    for (auto value : data_part)
+    {
+        encoded += lut[value];
+    }
+    for (auto value : checksum)
+    {
+        encoded += lut[value];
+    }
+    return encoded;
+}
+
+inline char toLowercase(const char& c)
+{
+    return c + ('a' - 'A');
+}
+
+bool decode_base32(const std::string& in, std::string& hr_part, data_chunk& values)
+{
+    // Basic input check
+    if (in.length() < base32_min_length || in.length() > base32_max_length)
+    {
+        return false;
+    }
+    
+    // Find separator and check valid hr_part and data_part sizes
+    size_t pos_separator = in.rfind(separator);
+    if (pos_separator == std::string::npos 
+        || pos_separator < hr_part_min_length
+        || pos_separator > hr_part_max_length
+        || in.length() - (pos_separator + 1) < data_part_min_length)
+    {
+        return false;
+    }
+    
+    // Convert to a lowercase string and check for valid characters
+    bool uppercase = false;
+    bool lowercase = false;
+    std::string encoded;
+    encoded.reserve(in.length());
+    for (auto c : in)
+    {
+        if (c < 33 || c > 126)
+            return false;
+        if (c >= 'A' && c <= 'Z')
+        {
+            uppercase = true;
+            c = toLowercase(c);
+        }
+        else if (c >= 'a' && c <= 'z')
+        {
+            lowercase = true;
+        }
+        encoded += c;
+    }
+    // Must NOT accept mixed case strings
+    if (uppercase && lowercase)
+    {
+        return false;
+    }
+    
+    // Convert data part to values
+    const size_t data_part_length = encoded.size() - (pos_separator+1);
+    data_chunk data_part_values;
+    data_part_values.reserve(data_part_length);
+    for (auto i = pos_separator + 1; i < encoded.size(); ++i)
+    {
+        if (rev_lut[encoded[i]] == -1)
+            return false;
+        data_part_values.push_back(rev_lut[encoded[i]]);
+    }
```

Comment:
```cpp
for (const unsigned char c : encoded.substr(pos_separator + 1, encoded.npos))
{
    if (rev_lut[c] == -1)
        return false;
    data_part_values.push_back(rev_lut[c]);
}
```
Otherwise, you'd be indexing `rev_lut` with something of type `char`. Also, I find it more readable this way :wink: